File: python/chapter03/adventure_game/objects/person.py
import math

from chapter03.adventure_game import world
from chapter03.adventure_game.adventure_substrate.messaging import narrate, say, announce
from .bag import Bag
from .mobile_thing import MobileThing
from ..generics import move

BASE_HEALTH = 100


def is_person(obj):
    return isinstance(obj, Person)


class Person(MobileThing):
    def __init__(self, name: str, home):
        super().__init__(name, home)
        self.health = BASE_HEALTH
        self.bag = Bag("my-bag", holder=self)

        world.the_clock.register(self)

    def get_things(self):
        return self.bag.get_things()

    def things_here(self):
        return self.location.things_in_place()

    def people_here(self):
        people = self.location.people_in_place()
        people.remove(self)
        return people

    def vistas_here(self):
        return self.location.vistas

    def exits_here(self):
        return list(self.location.exits.values())

    def peoples_things(self):
        return [thing for person in self.people_here() for thing in person.get_things()]

    def take_thing(self, thing):
        move(thing, self.bag, self)

    def drop_thing(self, thing):
        move(thing, self.location, self)

    def enter_place(self):
        super().enter_place()
        narrate([self, "enters", self.location], self)
        people = self.people_here()
        if people:
            say(self, ["Hi", *people])

    def suffer(self, hits):
        say(self, ["Ouch!", hits, "hits is more than I want!"])
        self.health -= hits
        if self.health <= 0:
            self.die()

    def heal(self, points):
        if self.health >= BASE_HEALTH:
            return

        self.health += points
        if self.health >= BASE_HEALTH:
            self.health = BASE_HEALTH

        if self.health == BASE_HEALTH:
            say(self, ["I feel like new again!"])
        else:
            say(self, ["I feel better."])

    def die(self):
        for thing in self.get_things():
            self.drop_thing(thing)
        announce("An earth-shattering, soul-piercing scream is heard...")
        self.health = 0
        move(self, world.heaven, self)

    def resurrect(self, health):
        assert health > 0
        self.health = health
        move(self, self.origin, self)

    def clock_tick(self):
        super().clock_tick()
        heal_amount = math.ceil((BASE_HEALTH - self.health) * 0.1)
        self.heal(heal_amount)

File: rust/src/chapter03/adventure_game/objects/person.rs
use crate::chapter03::adventure_game::objects::mobile_thing::is_mobile_thing;
use crate::chapter03::adventure_game::objects::place::{
    get_exits, get_vistas, people_in_place, things_in_place,
};
use crate::chapter03::adventure_game::objects::thing::get_location;
use crate::chapter03::adventure_game::objects::{bag, mobile_thing};
use crate::chapter03::adventure_game::property_table::Properties;
use crate::chapter03::dynamic_type;
use crate::chapter03::dynamic_type::Obj;
use crate::chapter03::generic_procedures::predicate::declare_superset;
use crate::chapter03::DebugAny;

pub fn is_person(obj: &dyn DebugAny) -> bool {
    mobile_thing::is_mobile_thing(obj) && obj.has_property("health") && obj.has_property("bag")
}

pub fn make_person(name: impl ToString, location: Obj) -> Obj {
    let person = mobile_thing::make_mobile_thing(name, location.clone());
    person.set_raw_property("health", 3);
    person.set_property(
        "bag",
        bag::make_bag("my bag".to_string(), dynamic_type::obj(false)),
    );
    return person;
}

pub fn get_bag(person: &Obj) -> Obj {
    assert!(is_person(&**person));
    person.get_property("bag").unwrap()
}

pub fn things_here(person: &Obj) -> Vec<Obj> {
    assert!(is_person(&**person));
    things_in_place(&get_location(person))
}

pub fn people_here(person: &Obj) -> Vec<Obj> {
    assert!(is_person(&**person));
    let mut people = people_in_place(&get_location(person));
    let me = people.iter().position(|p| p == person).unwrap();
    people.remove(me);
    people
}

pub fn vistas_here(person: &Obj) -> Vec<Obj> {
    assert!(is_person(&**person));
    get_vistas(&get_location(person))
}

pub fn exits_here(person: &Obj) -> Vec<Obj> {
    assert!(is_person(&**person));
    get_exits(&get_location(person))
}

pub fn install_generic_procedure_handlers() {
    declare_superset(is_person, is_mobile_thing);
}
